### PR TITLE
New py of making an order

### PR DESCRIPTION
**Pull request by Richymaestro**
Made a new version of the order.py, in style of OOP Python. Let me know if it works and if you have remarks!

### CODE REVIEW

File: src/CowOrder.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from hexbytes import HexBytes
+import time
+
+class Fee: 
+    def __init__(self, base_url, feeAddress):
+        self.base_url = base_url
+        self.feeAddress = feeAddress
+    
+    def setFee(self):
+        requestFee = requests.get(base_url + feeAddress + sellToken + "&buyToken=" + buyToken + "&amount=" +sellAmount + "&kind=" + kind)
```
Comment (bh2smith) [inline src/CowOrder.py]: Generally speaking, I am not a fan of raw string concatenation. I think the code would be much more readable and reliable if we were to use python's string formatting with variable names as follow

```py
url = "{baseUrl}fee?sellToken={sellToken}&buyToken={buyToken}&amount={amount}&kind={kind}".format(
    baseUrl=base_url, 
    sellToken=self.sell_token, 
    buyToken=self.buy_token,
    amount=self.sell_amount,
    kind=self.kind
)
```

File: src/CowOrder.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        self.signature = Signature.getSig(self)
+        self.signingScheme = signingScheme
+        self.sellTokenBalance: String() = sellTokenBalance
+        self.buyTokenBalance: String() = buyTokenBalance 
+    
+    def toJson(self):
+        jsonorder = json.dumps(self, default=lambda obj: obj.__dict__, indent=4)
+        self.orders = orders
+        self.base_url = base_url
+        return requests.post(self.base_url + self.orders + jsonorder)
```
Comment (bh2smith) [inline src/CowOrder.py]: If we look at the post made [here]() in the original script 

```
r = requests.post(base_url + "orders", json=j)
```
this is a string concatenation of the base url with the string "orders" followed by the json data of the order

```suggestion
        return requests.post("{base_url}orders".format(base_url), json=jsonorder)
```

File: src/CowOrder.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        self.appData: Bytes(32) = appData
+        self.feeAmount: Uint(256) = str(Fee.setFee(self))
+        self.partiallyFillable: Boolean() = partiallyFillable
+        self.signature = Signature.getSig(self)
+        self.signingScheme = signingScheme
+        self.sellTokenBalance: String() = sellTokenBalance
+        self.buyTokenBalance: String() = buyTokenBalance 
+    
+    def toJson(self):
+        jsonorder = json.dumps(self, default=lambda obj: obj.__dict__, indent=4)
```
Comment (bh2smith) [inline src/CowOrder.py]: This method is called "toJson" so we would expect that it merely returns the jsonorder declared in the first line here. However, this method appears to be posting an order.

I would suggest 
```suggestion
    def to_json(self):
        return json.dumps(self, default=lambda obj: obj.__dict__, indent=4)
```

followed by another method `post_order` which does something like
```py
return requests.post("{base_url}orders".format(base_url), json=self.to_json())
```

File: src/CowOrder.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        self.buyToken: Address() = buyToken
+        self.sellAmount: Uint(256) = sellAmount
+        self.kind: String() = kind
+        self.receiver: Address() = receiver
+        self.buyAmount: Uint(256) = buyAmount
+        self.validTo: Uint(32) = validTo
+        self.appData: Bytes(32) = appData
+        self.feeAmount: Uint(256) = str(Fee.setFee(self))
+        self.partiallyFillable: Boolean() = partiallyFillable
+        self.signature = Signature.getSig(self)
```
Comment (bh2smith) [inline src/CowOrder.py:60]: While this is a pretty good idea we may also want to update certain parameters of an order (like the fee amount) if it changes. In this case I might also include a method `update_fee` which fetches a new fee quote and updates the value. Along with this, then we want a method like "Order.sign()" which fills the signature filed based on all the current values.

**Review by bh2smith — CHANGES_REQUESTED** [review]
Hello @Richymaestro and thanks for your very nice looking contribution! Very glad to have you interested in the project.

Everything is looking very good so far, but there are a few places here which aren't quite correct and others that could be improved.

Given that this is quite involved (i.e. contains the introduction of several classes and methods). I might also suggest introducing them each as separate PRs. It is, however pretty clear that all the classes are quite interconnected, so we could allow them for now pending some cleanup and unit testing afterwards.

Each class should be in its own file
```
fee.py
signature.py
order.py
```

Another thing that would probably be more appropriate is the include your changes directly in the existing file. Once you can actually manage to post an order with the content of this code, it would be really good if you could include a "Test Plan" in the PR description demonstrating that an order was placed using it.

**Comment by Richymaestro** [comment]
Thanks for your comments, I will add them tomorrow!

File: src/CowOrder.py (hunk 1 above)
Comment (Richymaestro) [reply src/CowOrder.py]: Yes, I agree with that, will change this one

File: src/CowOrder.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import time
+
+class Fee: 
+    def __init__(self, sellToken, buyToken, sellAmount, kind):
+        self.sellToken = sellToken
+        self.buyToken = buyToken
+        self.sellAmount = sellAmount
+        self.kind = kind
+    
+    def setFee(self):
```
Comment (bh2smith) [inline src/CowOrder.py:17]: This method would be more appropriately named as get_fee since it fetches it and returns it.

**Comment by Richymaestro** [comment]
Made a new PR with updated code, will close this one

**Pull request closed without merging**